src/netrc.c

```diff
@@ -37,78 +37,127 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <stdbool.h>
-#include "axel.h"
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
 #include "netrc.h"
 
-enum lookup_state {
-	NOTHING,
-	DEFAULT,
-	MACHINE,
-	HOST_ENTRY,
-	LOGIN,
-	PASSWORD
-};
-
-void
-netrc_parse(conn_t *conn)
+#define REL_NETRC	"/.netrc"
+#define IS_DELIM(x)	(*x == ' ' || *x == '\t' || *x == '\n')
+
+static int fd;
+static size_t sz;
+static char *p;
+static char *s_addr;
+static char *e_addr;
+
+static size_t file_size(void)
+{
+	struct stat st;
+	fstat(fd, &st);
+	return st.st_size;
+}
+
+static int netrc_open(const char * const filename)
 {
-	FILE *fp;
-	bool found = false;
+	int ret;
+	char *path;
+	char *netrc;
 	char *home;
-	char *tok, *ntok;
-	char line[MAX_STRING];
-	char const *delims = " \t\n";
-	enum lookup_state state = NOTHING;
-
-	if (conn->conf->netrc_filename[0] == '\0') {
-		home = getenv("HOME");
-		snprintf(conn->conf->netrc_filename, MAX_STRING, "%s/.netrc", home);
+
+	ret = 0;
+	if (filename && filename[0] != '\0') {
+		path = strdup(filename);
+	} else {
+		if ((netrc = getenv("NETRC"))) {
+			path = strdup(netrc);
+		} else if ((home = getenv("HOME"))) {
+			size_t i = strlen(home) + strlen(REL_NETRC) + 1;
+			path = malloc(i);
+			snprintf(path, i, "%s%s", home, REL_NETRC);
+		} else {
+			return ret;
+		}
 	}
-	if ((fp = fopen(conn->conf->netrc_filename, "r")) == NULL)
-		return;
-	while (!found && fgets(line, MAX_STRING, fp)) {
-		tok = strtok_r(line, delims, &ntok);
-		if (tok && *tok == '#')
-			continue;
-		while (tok != NULL) {
-			switch (state) {
-			case NOTHING:
-				if (!strcmp("machine", tok)) {
-					state = MACHINE;
-				} else if (!strcmp("default", tok)) {
-					state = DEFAULT;
-				}
-				break;
-			case MACHINE:
-				if (!strcmp(conn->host, tok)) {
-					state = HOST_ENTRY;
-				} else {
-					state = NOTHING;
-				}
-				break;
-			case DEFAULT:
-				state = HOST_ENTRY;
-				// fall through
-			case HOST_ENTRY:
-				if (!strcmp("login", tok)) {
-					state = LOGIN;
-				} else if (!strcmp("password", tok)) {
-					state = PASSWORD;
-				}
-				break;
-			case LOGIN:
-				strcpy(conn->user, tok);
-				state = HOST_ENTRY;
-				break;
-			case PASSWORD:
-				strcpy(conn->pass, tok);
-				found = true;
-				break;
-			default:
+	if ((fd = open(path, O_RDONLY, 0)) == -1) {
+		ret = 0;
+	} else {
+		sz = file_size();
+		s_addr = mmap(NULL, sz, PROT_READ, MAP_PRIVATE | MAP_POPULATE, fd, 0);
+		e_addr = s_addr + sz;
+		ret = 1;
+	}
+	free(path);
+	return ret;
+}
+
+static void netrc_close(void)
+{
+	munmap(s_addr, sz);
+	close(fd);
+}
+
+static int next_token(char **t)
+{
+	char *q;
+	int len;
+
+	if (!p)
+		p = s_addr;
+	while (IS_DELIM(p) && p < e_addr)
+		p++;
+	q = p;
+	while (!IS_DELIM(q) && q < e_addr)
```

Comment:
Wouldn't it make the function too generic? I mean, we already know the tokens in `.netrc` file are separated only by spaces, tabs or new-lines, which are checked by the macro `IS_DELIM(x)`. I'm failing to see the benefits of the changes in case of this static aux function.

Reply:
Well, every new functionality will likely influence refactoring of other parts of the program, so better be generic, it's not going to hurt footprint nor performance (in fact `strcspn` is generally faster than your `IS_DELIM` macro, believe it or not...).

Reply:
Also, if you use some interface that looks like the standard one, chances are you'll be able to take that code and modify it with much less work.